docs/shared/recipes/generators/register-sync-generator.md

```diff
@@ -0,0 +1,110 @@
+# Register a Sync Generator
+
+Sync generators are intended to keep applicable files in sync with your project graph before a task is run. From a technical perspective, a sync generator is no different from any other generator, but it has some additional performance considerations and needs to be registered in a particular way.
+
+## Create a Sync Generator
+
+Create a sync generator the same way you would [create any generator](/extending-nx/recipes/local-generators):
+
+```shell
+nx g my-sync-generator --directory=tools/my-plugin/src/generators/my-sync-generator
+```
+
+A sync generator should be able to run without any required options, so update the schema accordingly:
+
+```jsonc {% fileName="tools/my-plugin/src/generators/my-sync-generator/schema.json" %}
+{
+  "$schema": "https://json-schema.org/schema",
+  "$id": "MySyncGenerator",
+  "title": "",
+  "type": "object",
+  "properties": {},
+  "required": []
+}
+```
+
+Sync generators can optionally return an `outOfSyncMessage` to display to users when the sync generator needs to be run.
+
+```ts {% fileName="tools/my-plugin/src/generators/my-sync-generator/my-sync-generator.ts" %}
+import { Tree } from '@nx/devkit';
+import { SyncGeneratorResult } from 'nx/src/utils/sync-generators';
+
+export async function mySyncGenerator(
+  tree: Tree
+): Promise<SyncGeneratorResult> {
+  if (
+    !tree.exists('/legal-message.txt') ||
+    tree.read('/legal-message.txt').toString() !==
+      'This is an important legal message.'
+  ) {
+    tree.write('/legal-message.txt', 'This is an important legal message.');
+  }
+  return {
+    outOfSyncMessage: 'The legal-message.txt file needs to be created',
+  };
+}
+
+export default mySyncGenerator;
+```
+
+## Performance Considerations
```

Comment:
I think that a recipe about creating sync generators should also suggest that sync generator authors be informative when throwing an error or returning an `outOfSyncMessage.` They should provide guidance to the user on how to address any error.

This is not a perf-related suggestion but an important DX suggestion. Sync generators can block tasks, so providing good feedback is important.